### PR TITLE
ODATA-1532: OData-Aggr WD05 sections 3.10.2.2 and 3.22

### PR DESCRIPTION
**Pull request by HeikoTheissen**

### CODE REVIEW

File: abnf/odata-aggregation-testcases.yaml [hunk 1]
```diff
@@ -752,7 +752,7 @@ TestCases:
 
   - Name: hierarchy transformations - traverse
     Rule: queryOptions
-    Input: $apply=traverse(SalesOrgHierarchy,preorder)
+    Input: $apply=traverse($root/SalesOrganizations,SalesOrgHierarchy,ID,preorder)
```
Comment (ralfhandl) [inline abnf/odata-aggregation-testcases.yaml:755]: I would have expected to surround the first three with parens to make them a triple, just as in `rollup()`

Comment (HeikoTheissen) [reply abnf/odata-aggregation-testcases.yaml:755]: No, it's not like in `rollup` but like in `ancestors`.

**Review by ralfhandl — APPROVED** [review]

File: abnf/odata-aggregation-abnf.txt [hunk 2]
```diff
@@ -230,12 +230,11 @@ searchTrafo     = %s"search" OPEN BWS searchExpr BWS CLOSE
...
 groupbyTrafo    = %s"groupby" OPEN BWS groupbyList [ BWS COMMA BWS applyExpr ] BWS CLOSE
 groupbyList     = OPEN BWS groupbyElement *( BWS COMMA BWS groupbyElement ) BWS CLOSE
-groupbyElement  = groupingProperty / rollup / rollupAll
+groupbyElement  = groupingProperty / rollup / rollupDesc
 
-rollup            = %s"rollup"    OPEN BWS ( rollupUnnamedHier / rollupNamedHier ) BWS CLOSE
-rollupAll         = %s"rollupall" OPEN BWS ( rollupUnnamedHier / rollupNamedHier ) BWS CLOSE
-rollupUnnamedHier = groupingProperty 1*( BWS COMMA BWS groupingProperty )
-rollupNamedHier   = recHierQualifier
+rollup          = ( %s"rollupall" / %s"rollup" )
```
Comment (GeraldKrause) [inline abnf/odata-aggregation-abnf.txt:235]: 1. This rule now defines rolling up a leveled hierarchy. I suggest to reflect this in the name, e.g. rollupLevels. 
2. This rule must include named leveled hierarchies, which are still missing here.

**Review by GeraldKrause — CHANGES_REQUESTED** [review]

File: abnf/odata-aggregation-abnf.txt (hunk 2 above)
Comment (HeikoTheissen) [reply abnf/odata-aggregation-abnf.txt:235]: 1. Wouldn't a name change `rollup` -> `rollupLevels` be incompatible?
2. The named leveled hierarchies are in line 236:
https://github.com/oasis-tcs/odata-abnf/blob/0ce3356e5eef203bae4212087daa8abcd1b3af0a/abnf/odata-aggregation-abnf.txt#L235-L236

**Review by GeraldKrause — APPROVED** [review]

File: abnf/odata-aggregation-abnf.txt [hunk 3]
```diff
@@ -230,12 +230,12 @@ searchTrafo     = %s"search" OPEN BWS searchExpr BWS CLOSE
 
 groupbyTrafo    = %s"groupby" OPEN BWS groupbyList [ BWS COMMA BWS applyExpr ] BWS CLOSE
 groupbyList     = OPEN BWS groupbyElement *( BWS COMMA BWS groupbyElement ) BWS CLOSE
-groupbyElement  = groupingProperty / rollup / rollupAll
+groupbyElement  = groupingProperty / rollupLevels / rollupDesc
 
-rollup            = %s"rollup"    OPEN BWS ( rollupUnnamedHier / rollupNamedHier ) BWS CLOSE
-rollupAll         = %s"rollupall" OPEN BWS ( rollupUnnamedHier / rollupNamedHier ) BWS CLOSE
+rollupLevels      = ( %s"rollupall" / %s"rollup" ) OPEN BWS ( rollupUnnamedHier / rollupNamedHier ) BWS CLOSE
+rollupDesc        = %s"rollupdescendants" OPEN BWS recHierReference BWS CLOSE
```
[the remark this answers is not in the document]
Comment (GeraldKrause) [reply abnf/odata-aggregation-abnf.txt]: Another idea:

groupby((rolluprecursiveall(...)), aggregate(<expr>))

= 

concat(
      groupby((rolluprecursive(...)), aggregate(<expr>)),
      aggregate(<expr>))

We could get rid of the ...all operators and instead provide an example for grand totals in section 7.

[the remark this answers is not in the document]
Comment (GeraldKrause) [reply abnf/odata-aggregation-abnf.txt]: 2022-07-21: Agree to go without ...all and provide examples in the spec text

**Pull request merged**